src/spyglass/decoding/v1/clusterless.py

```diff
@@ -415,3 +423,93 @@ def load_spike_data(key, filter_by_interval=True):
             new_waveform_features.append(elec_waveform_features[is_in_interval])
 
         return new_spike_times, new_waveform_features
+
+    @classmethod
+    def get_spike_indicator(cls, key, time):
+        time = np.asarray(time)
+        min_time, max_time = time[[0, -1]]
+        spike_times = cls.load_spike_data(key)[0]
+        spike_indicator = np.zeros((len(time), len(spike_times)))
+
+        for ind, times in enumerate(spike_times):
+            times = times[np.logical_and(times >= min_time, times <= max_time)]
+            spike_indicator[:, ind] = np.bincount(
+                np.digitize(times, time[1:-1]),
+                minlength=time.shape[0],
+            )
+
+        return spike_indicator
+
+    @classmethod
+    def get_firing_rate(cls, key, time, multiunit=False):
+        spike_indicator = cls.get_spike_indicator(key, time)
+        if spike_indicator.ndim == 1:
+            spike_indicator = spike_indicator[:, np.newaxis]
+
+        sampling_frequency = 1 / np.median(np.diff(time))
+
+        if multiunit:
+            spike_indicator = spike_indicator.sum(axis=1, keepdims=True)
+        return np.stack(
+            [
+                get_multiunit_population_firing_rate(
+                    indicator[:, np.newaxis], sampling_frequency
+                )
+                for indicator in spike_indicator.T
+            ],
+            axis=1,
+        )
+
+    def get_ahead_behind_distance(self):
+        # TODO: allow specification of specific time interval
+        # TODO: allow specification of track graph
+        # TODO: Handle decode intervals, store in table
+
+        classifier = self.load_model()
+        results = self.load_results()
+        posterior = results.acausal_posterior.unstack("state_bins").sum("state")
+
+        if classifier.environments[0].track_graph is not None:
```

Comment:
This might protect against issues with novel environment structures
 ```suggestion
        if getattr(classifier.environments[0], "track_graph") is not None:
```